drivers/net/ethernet/renesas/rswitch.c

```diff
@@ -2368,14 +2383,115 @@ static int rswitch_modify_l3fwd(struct l3_ipv4_fwd_param *param, bool delete)
 	else
 		rs_write32(param->dv, priv->addr + FWLTHTL9);
 
-	return rswitch_reg_wait(priv->addr, FWLTHTLR, LTHTL, 0);
+	res = rswitch_reg_wait(priv->addr, FWLTHTLR, LTHTL, 0);
+	if (res)
+		return res;
+
+	res = rs_read32(priv->addr + FWLTHTLR);
+	collision_num = L3_LEARN_COLLISSION_NUM(res);
+	if ((collision_num > priv->max_collisions) && !delete)
+		return -EAGAIN;
+
+	return 0;
 }
 
 int rswitch_add_l3fwd(struct l3_ipv4_fwd_param *param)
 {
 	return rswitch_modify_l3fwd(param, false);
 }
 
+/* Should be called with rswitch_priv->ipv4_forward_lock taken */
+static int rswitch_restore_l3_table(struct rswitch_private *priv)
+{
+	struct list_head *cur, *cur_param_list;
+	struct rswitch_ipv4_route *routing_list;
+	struct l3_ipv4_fwd_param_list *param_list;
+#if IS_ENABLED(CONFIG_IP_MROUTE)
+	struct rswitch_ipv4_multi_route *multi_route;
+#endif
+	struct rswitch_device *rdev;
+	int rc = 0;
+
+	read_lock(&priv->rdev_list_lock);
+	list_for_each_entry(rdev, &priv->rdev_list, list) {
+		rc = rswitch_restore_tc_l3_table(rdev);
+		if (rc)
+			goto unlock;
+		list_for_each(cur, &rdev->routing_list) {
+			routing_list = list_entry(cur, struct rswitch_ipv4_route, list);
+			list_for_each(cur_param_list, &routing_list->param_list) {
+				param_list =
+					list_entry(cur_param_list,
+						   struct l3_ipv4_fwd_param_list,
+						   list);
+				rc = rswitch_add_l3fwd(param_list->param);
+				if (rc)
+					goto unlock;
+			}
+		}
+
+#if IS_ENABLED(CONFIG_IP_MROUTE)
+		list_for_each(cur, &rdev->mult_routing_list) {
+			multi_route = list_entry(cur, struct rswitch_ipv4_multi_route, list);
+			rc = rswitch_add_l3fwd(&multi_route->params[0]);
+			if (rc)
+				goto unlock;
+			rc = rswitch_add_l3fwd(&multi_route->params[1]);
+			if (rc)
+				goto unlock;
+		}
+#endif
+	}
+
+unlock:
+	read_unlock(&priv->rdev_list_lock);
+
+	return rc;
+}
+
+/* This function is preferred to use instead of rswitch_add_l3fwd in
+ * case of adding L3 streaming entry. It checks rswitch_add_l3fwd
+ * result and if EAGAIN is returned the function adjusts equation
+ * to reduce the collision number. There is no reason to use it
+ * for perfect filter because in this case, collisions won't happenned.
+ * Should be called with rswitch_priv->ipv4_forward_lock taken.
+ */
+int rswitch_add_l3fwd_adjust_hash(struct l3_ipv4_fwd_param *param)
+{
+	struct rswitch_private *priv = param->priv;
+	int rc;
+	u16 original_equation = priv->hash_equation;
+
+	do {
+		rc = rswitch_add_l3fwd(param);
+		if (rc == -EAGAIN) {
+			do {
+				priv->hash_equation++;
+				if (priv->hash_equation > FWLTHHC_LTHHE_MASK)
```

Comment:
Is the `FWLTHHC_LTHHE_MASK` correct constant to be checked against? Should it be  `FWLTHHC_LTHHE_MAX`?

Reply:
Oh, yes, I wrote "MASK" because in this case, it has the same sense (because configuration values are placed at the lower bits of the register). Nevertheless, for sure, it's more clear to use "MAX" suffix in this case, so I changed it.